File: wsnet/batched.go
package wsnet

import (
	"net"
	"sync"
	"time"

	"github.com/gorilla/websocket"
)

// BatchedOptions configures batched write and stream read behavior (used on both ends).
type BatchedOptions struct {
	// FlushInterval is the maximum time to hold data before sending a WebSocket
	// message. A single frame is sent within this long even if the batch never fills.
	// Zero uses DefaultFlushInterval (2ms).
	FlushInterval time.Duration
	// MaxBatchBytes is the maximum buffered bytes before flushing. Larger values
	// improve throughput under load; smaller values reduce latency.
	// Zero uses DefaultMaxBatchBytes (32 * 1500, i.e. ~32 MTU-sized packets).
	MaxBatchBytes int
}

const (
	DefaultFlushInterval = 2 * time.Millisecond
	DefaultMaxBatchBytes = 32 * 1500 // 32 * typical MTU
)

// BatchedConn wraps a WebSocket connection for use on both server and client:
// - Write: buffered; flushes when buffer reaches MaxBatchBytes or after FlushInterval (single frames still go out promptly).
// - Read: stream; each WebSocket message may contain many bytes; Read() drains the current message and fetches the next when empty.
// Use the same wrapper on both ends so both sides batch sends and consume batched receives as a stream.
func BatchedConn(conn *websocket.Conn, opts BatchedOptions) net.Conn {
	if opts.FlushInterval == 0 {
		opts.FlushInterval = DefaultFlushInterval
	}
	if opts.MaxBatchBytes == 0 {
		opts.MaxBatchBytes = DefaultMaxBatchBytes
	}
	c := &batchedConn{
		conn: conn,
		opts: opts,
		buf:  make([]byte, 0, opts.MaxBatchBytes*2), // avoid realloc for typical batch
	}
	return c
}

type batchedConn struct {
	conn    *websocket.Conn
	opts    BatchedOptions
	mu      sync.Mutex
	buf     []byte
	timer   *time.Timer
	closed  bool
	readMu  sync.Mutex
	readBuf []byte
	readOff int
}

func (c *batchedConn) Write(p []byte) (n int, err error) {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed {
		return 0, net.ErrClosed
	}
	c.buf = append(c.buf, p...)
	n = len(p)
	if len(c.buf) >= c.opts.MaxBatchBytes {
		return n, c.flushLocked()
	}
	c.armTimerLocked()
	return n, nil
}

func (c *batchedConn) armTimerLocked() {
	if c.timer != nil {
		c.timer.Stop()
		c.timer = nil
	}
	c.timer = time.AfterFunc(c.opts.FlushInterval, c.onFlush)
}

func (c *batchedConn) onFlush() {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.timer = nil
	if len(c.buf) > 0 {
		_ = c.flushLocked()
	}
}

func (c *batchedConn) flushLocked() error {
	if len(c.buf) == 0 {
		return nil
	}
	err := c.conn.WriteMessage(websocket.BinaryMessage, c.buf)
	c.buf = c.buf[:0]
	return err
}

func (c *batchedConn) Read(p []byte) (n int, err error) {
	c.readMu.Lock()
	defer c.readMu.Unlock()
	for len(p) > 0 {
		if c.readOff >= len(c.readBuf) {
			_, c.readBuf, err = c.conn.ReadMessage()
			if err != nil {
				return n, err
			}
			c.readOff = 0
		}
		nc := copy(p, c.readBuf[c.readOff:])
		c.readOff += nc
		n += nc
		p = p[nc:]
	}
	return n, nil
}

func (c *batchedConn) Close() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	if c.closed {
		return nil
	}
	c.closed = true
	if c.timer != nil {
		c.timer.Stop()
		c.timer = nil
	}
	_ = c.flushLocked()
	return c.conn.Close()
}

func (c *batchedConn) LocalAddr() net.Addr  { return c.conn.LocalAddr() }
func (c *batchedConn) RemoteAddr() net.Addr { return c.conn.RemoteAddr() }
func (c *batchedConn) SetDeadline(t time.Time) error {
	_ = c.conn.SetReadDeadline(t)
	_ = c.conn.SetWriteDeadline(t)
	return nil
}
func (c *batchedConn) SetReadDeadline(t time.Time) error  { return c.conn.SetReadDeadline(t) }
func (c *batchedConn) SetWriteDeadline(t time.Time) error { return c.conn.SetWriteDeadline(t) }

var _ net.Conn = (*batchedConn)(nil)

File: wsnet/batched.js
/**
 * BatchedWebSocket – browser WebSocket replacement that matches wsnet.BatchedConn behavior.
 *
 * - send(data): buffers and flushes when buffer reaches maxBatchBytes or after flushInterval
 *   (single frames still go out within flushInterval).
 * - read(buffer): reads received bytes into the given Uint8Array; resolves with bytes read.
 *   Incoming WebSocket messages are drained as a stream (same as Go BatchedConn Read).
 *
 * Use the same options on both ends (this client and the Go server with wsnet.BatchedConn).
 */

const DEFAULT_FLUSH_INTERVAL_MS = 2;
const DEFAULT_MAX_BATCH_BYTES = 32 * 1500;

export class BatchedWebSocket {
  /**
   * @param {string|WebSocket} urlOrWs - WebSocket URL to connect, or an existing WebSocket to wrap
   * @param {Object} [opts]
   * @param {number} [opts.flushIntervalMs] - Max ms to hold data before sending (default 2)
   * @param {number} [opts.maxBatchBytes] - Flush when buffered bytes >= this (default 32*1500)
   */
  constructor(urlOrWs, opts = {}) {
    this.flushIntervalMs = opts.flushIntervalMs ?? DEFAULT_FLUSH_INTERVAL_MS;
    this.maxBatchBytes = opts.maxBatchBytes ?? DEFAULT_MAX_BATCH_BYTES;

    this._ws = typeof urlOrWs === 'string' ? new WebSocket(urlOrWs) : urlOrWs;
    this._writeBuf = [];
    this._writeLen = 0;
    this._flushTimer = null;

    /** @type {Uint8Array[]} - received chunks not yet consumed by read() */
    this._readChunks = [];
    this._readChunkOff = 0;
    /** @type {Array<{resolve: (n: number) => void, buffer: Uint8Array}>} */
    this._readWaiters = [];

    this._ws.binaryType = 'arraybuffer';
    this._ws.onmessage = (ev) => {
      const data = ev.data;
      const view = data instanceof ArrayBuffer ? new Uint8Array(data) : new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
      this._readChunks.push(view);
      this._drainReadWaiters();
    };
    this._userOnclose = null;
    this._ws.onclose = (ev) => {
      while (this._readWaiters.length > 0) {
        const { resolve } = this._readWaiters.shift();
        resolve(0);
      }
      if (this._userOnclose) this._userOnclose(ev);
    };
  }

  get readyState() {
    return this._ws.readyState;
  }

  get url() {
    return this._ws.url;
  }

  set onopen(fn) {
    this._ws.onopen = fn;
  }
  set onclose(fn) {
    this._userOnclose = fn;
  }
  set onerror(fn) {
    this._ws.onerror = fn;
  }

  /**
   * Send data; may be buffered and flushed on timer or when batch is full.
   * @param {ArrayBuffer|ArrayBufferView|Uint8Array} data
   */
  send(data) {
    const view = data instanceof ArrayBuffer ? new Uint8Array(data) : new Uint8Array(data.buffer, data.byteOffset, data.byteLength);
    this._writeBuf.push(view);
    this._writeLen += view.length;

    if (this._writeLen >= this.maxBatchBytes) {
      this._flush();
    } else {
      this._armFlushTimer();
    }
  }

  _armFlushTimer() {
    if (this._flushTimer != null) return;
    this._flushTimer = setTimeout(() => {
      this._flushTimer = null;
      this._flush();
    }, this.flushIntervalMs);
  }

  _flush() {
    if (this._flushTimer != null) {
      clearTimeout(this._flushTimer);
      this._flushTimer = null;
    }
    if (this._writeBuf.length === 0) return;

    const total = this._writeLen;
    const out = new Uint8Array(total);
    let off = 0;
    for (const chunk of this._writeBuf) {
      out.set(chunk, off);
      off += chunk.length;
    }
    this._writeBuf = [];
    this._writeLen = 0;

    if (this._ws.readyState === WebSocket.OPEN) {
      this._ws.send(out.buffer);
    }
  }

  /**
   * Read up to buffer.length bytes from the received stream; resolves with bytes read.
   * If no data is available, waits for the next WebSocket message.
   * @param {Uint8Array} buffer
   * @returns {Promise<number>}
   */
  read(buffer) {
    return new Promise((resolve) => {
      const n = this._copyOut(buffer);
      if (n > 0) {
        resolve(n);
        return;
      }
      this._readWaiters.push({ resolve, buffer });
    });
  }

  _copyOut(buffer) {
    let total = 0;
    while (total < buffer.length && this._readChunks.length > 0) {
      const chunk = this._readChunks[0];
      const rest = chunk.length - this._readChunkOff;
      const n = Math.min(rest, buffer.length - total);
      buffer.set(chunk.subarray(this._readChunkOff, this._readChunkOff + n), total);
      total += n;
      this._readChunkOff += n;
      if (this._readChunkOff >= chunk.length) {
        this._readChunks.shift();
        this._readChunkOff = 0;
      }
    }
    return total;
  }

  _drainReadWaiters() {
    while (this._readWaiters.length > 0 && this._readChunks.length > 0) {
      const { resolve, buffer } = this._readWaiters.shift();
      const n = this._copyOut(buffer);
      resolve(n);
    }
  }

  /**
   * Flush any buffered send data and close the WebSocket.
   */
  close(code, reason) {
    this._flush();
    this._ws.close(code, reason);
  }
}

export default BatchedWebSocket;
